store/pg_engine.go

```diff
@@ -373,23 +376,143 @@ func migrate(ctx context.Context, d dbdriver.Driver, curVer *semver.Version, mod
+			return true
+		}
+	}
+	return false
+}
+
+type devMigration struct {
+	filename  string
+	version   string
+	statement string
+}
+
+func getDevMigrations() ([]devMigration, error) {
+	names, err := fs.Glob(migrationFS, fmt.Sprintf("migration/%s/*", common.ReleaseModeDev))
+	if err != nil {
+		return nil, err
 	}
 
-	return retVersion, nil
+	var devMigrations []devMigration
+	for _, name := range names {
+		baseName := filepath.Base(name)
+		if baseName == latestSchemaFile || baseName == latestDataFile {
+			continue
+		}
+
+		parts := strings.Split(filepath.Base(name), "__")
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("invalid migration file name %q", name)
+		}
+		buf, err := fs.ReadFile(migrationFS, name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read dev migration file %q, error %w", name, err)
+		}
+
+		devMigrations = append(devMigrations, devMigration{
+			filename:  name,
+			version:   parts[0],
+			statement: string(buf),
+		})
+	}
+	sort.Slice(devMigrations, func(i, j int) bool {
+		return devMigrations[i].version < devMigrations[j].version
+	})
+	return devMigrations, nil
 }
 
-func getCutoffVersion(releaseMode common.ReleaseMode) (semver.Version, error) {
-	minorPathPrefix := fmt.Sprintf("migration/%s/*", releaseMode)
+func migrateDev(ctx context.Context, d dbdriver.Driver, serverVersion, databaseName string, cutoffSchemaVersion semver.Version, histories []*dbdriver.MigrationHistory, l *zap.Logger) error {
+	devMigrations, err := getDevMigrations()
+	if err != nil {
+		return err
+	}
+
+	// Skip migrations that are already applied.
+	var migrations []devMigration
+	for _, m := range devMigrations {
+		if migrationExists(m.statement, histories) {
+			l.Info(fmt.Sprintf("Skip migrating dev migration file %s that's already migrated.", m.filename))
+		} else {
+			migrations = append(migrations, m)
+		}
+	}
+	if len(migrations) == 0 {
+		l.Info("Skip dev mode migration; no new version.")
+		return nil
+	}
+
+	for _, m := range migrations {
+		if migrationExists(m.statement, histories) {
+			l.Info(fmt.Sprintf("Skip migrating dev migration file %s that's already migrated.", m.filename))
+			continue
+		}
+		l.Info(fmt.Sprintf("Migrating dev %s...", m.filename))
+		if _, _, err := d.ExecuteMigration(
+			ctx,
+			&dbdriver.MigrationInfo{
+				ReleaseVersion: serverVersion,
+				Version:        m.version,
+				Namespace:      databaseName,
+				Database:       databaseName,
+				Environment:    "", /* unused in execute migration */
+				Source:         dbdriver.LIBRARY,
+				Type:           dbdriver.Migrate,
+				Description:    fmt.Sprintf("Migrate version %s server version %s with files %s.", m.version, serverVersion, m.filename),
+			},
+			m.statement,
+		); err != nil {
+			return fmt.Errorf("failed to migrate schema version %q, error: %v", m.version, err)
+		}
+	}
+	// Establish a baseline to restart the migration with semantic version.
+	if _, _, err := d.ExecuteMigration(
```

Comment:
Yes, since we use different migration strategies (non-semantic / semantic), we should do a baseline going from non-semantic to semantic so we know the latest (semantic version) when the program restarts next time. Otherwise, we have to traverse all migration history to find the latest semantic version.

Reply:
Understand this now.

The reason is getLatestVersion requires the last migration history to be semantic version. Actually, before this PR, we assume all migration history use semantic version.

I have another thought, what if we use semantic version for the dev migration as well?

If the cutoff version is 1.1.0, the dev migration file is 2020011311024__foo.sql the version is 1.1.0-dev2020011311024

Reply:
Understand this now.

The reason is getLatestVersion requires the last migration history to be semantic version. Actually, before this PR, we assume all migration history use semantic version (we probably need to comment that assumption there)

I have another thought, what if we use semantic version for the dev migration as well?

If the cutoff version is 1.1.0, the dev migration file is 2020011311024__foo.sql the version is 1.1.0-dev2020011311024

Reply:
That sounds good. I used an extra dash "dev-2020011311024" as the suffix.